### DISCUSSION TITLE
Create Test Cases for Interactions

### DISCUSSION THREAD

**Discussion by civilCornball** [discord-php/DiscordPHP#1234] [Q&A] [answered]
Testing interactions manually can get tedious, and often you'll end up forgetting one of them.

How do you write or find code that would test your Interaction commands for you?

**Comment by valzargaming** [comment, marked as answer]
For the command listeners, you're probably best off writing typecasted returns and letting your IDE guide you through it as it would throw errors to warn you if your command never had a proper return. If you check that all variables are properly set, your own methods returned properly, etc. you shouldn't have issues.

```php
$this->listenCommand('ping', function (Interaction $interaction): PromiseInterface
{
    return $interaction->respondWithMessage(MessageBuilder::new()->setContent('Pong!'));
});
```

As for validating the Command structure for `$commands->save($command)`, most people are fine with using the CommandBuilder helper class, however I've had little luck with it. I've been writing my own series of helper classes to supplement using the various helper classes we already directly, but it's still a work in progress: https://github.com/Valgorithms/Civilizationbot/blob/main/src/Civ13/CommandServiceManager.php

**Reply by civilCornball** [reply]
> For the command listeners, you're probably best off writing typecasted returns

Could you explain? I'm not quite sure what you mean.

**Reply by valzargaming** [reply]
Each command listener takes a callback as the 2nd parameter. As long as it always returns with an instance of PromiseInterface (or ExtendedPromiseInterface) the code won't fail. If there's something else inside of your callback that's causing an issue, you should probably abstract it to a separate function and set up additional error handling.
